### ISSUE TITLE
multi-node training hangs occasionally

### ISSUE THREAD

**Issue by feiyulv** [apache/incubator-mxnet#4887]
@mli  The multi-node training still occasionally hangs with the newest mxnet version. Is there  anyone has occured the similar problem?

**Comment by weiliu89** [comment]
I am new to MXNet. I want to try to train inception-v3 to compare with TensorFlow (r1.0).

Environment:
4 Titan X Pascal with CUDNN 5.1
i7 5930 and 32G main memory
SSD to store data

Following tensorflow/models/inception, it takes about 8 days (750k iterations) to reach 76.95% on ImageNet (with batch_size = 128 ==> 32 per GPU). Training speed is about 160 images/sec. It takes about 8G per GPU and uses 16G main memory for caching data.

I also followed image-classification examples. With similar setting (i.e. 128 batch size), I noticed that MXNet is faster (~360 images/sec) and only needs 5.5G per GPU.

However, after a while, I observed a similar issue as @feiyulv, the training speed drops significantly:

INFO:root:Epoch[0] Batch [2500] Speed: 356.23 samples/sec Train-accuracy=0.011953
INFO:root:Epoch[0] Batch [2600] Speed: 357.26 samples/sec Train-accuracy=0.011406
INFO:root:Epoch[0] Batch [2700] Speed: 158.80 samples/sec Train-accuracy=0.012812
INFO:root:Epoch[0] Batch [2800] Speed: 41.29 samples/sec  Train-accuracy=0.014609
INFO:root:Epoch[0] Batch [2900] Speed: 40.90 samples/sec  Train-accuracy=0.014531

Then, after a while, it goes back normal again:

INFO:root:Epoch[0] Batch [4700] Speed: 40.96 samples/sec  Train-accuracy=0.039688
INFO:root:Epoch[0] Batch [4800] Speed: 40.59 samples/sec  Train-accuracy=0.037656
INFO:root:Epoch[0] Batch [4900] Speed: 75.93 samples/sec  Train-accuracy=0.041250
INFO:root:Epoch[0] Batch [5000] Speed: 367.78 samples/sec Train-accuracy=0.041328
INFO:root:Epoch[0] Batch [5100] Speed: 370.06 samples/sec Train-accuracy=0.041641

But it becomes worse again later...

I also used `test_io` option, and it says it can process about 450 images/sec. It looks like CPU and disk has the power to catch up the training speed.

It would be great if anyone can help on this.

**Comment by feiyulv** [comment]
@weiliu89 Our problems are different. I ran mxnet on multi-machine and the program hangs after several epochs. 
```
for i in range(10):
    python dmlc-submit --cluster mpi --num-servers 2 --num-workers 8  --hosts host  python train.py xxx
```
The program hangs when i > 1 occasionally. I observed that the process still exists after finishing last epoch  training. Would this cause the hang problem since they may use the same DMLC_PS_ROOT_PORT  @mli ? Do I need to kill the program manually after each mpi training?

**Comment by feiyulv** [comment]
This is the log when set PS_VERBOSE=1
```
2017-02-06 04:19:56,004 INFO Start 4 workers by mpirun
2017-02-06 04:19:56,022 INFO Start 4 servers by mpirun
[04:19:57] src/van.cc:75: Bind to role=scheduler, id=1, ip=172.20.7.4, port=9093, is_recovery=0
[04:19:57] src/van.cc:75: Bind to role=server, ip=11.11.72.4, port=44464, is_recovery=0
[04:19:57] src/van.cc:75: Bind to role=server, ip=11.11.72.4, port=44585, is_recovery=0
[04:19:57] src/van.cc:75: Bind to role=server, ip=11.11.72.4, port=44899, is_recovery=0
[04:19:57] src/van.cc:75: Bind to role=server, ip=11.11.72.4, port=56864, is_recovery=0
[04:19:57] src/van.cc:75: Bind to role=worker, ip=11.11.72.4, port=48209, is_recovery=0
[04:19:57] src/van.cc:75: Bind to role=worker, ip=11.11.72.4, port=34308, is_recovery=0
[04:19:57] src/van.cc:75: Bind to role=worker, ip=11.11.72.4, port=53854, is_recovery=0
[04:19:57] src/van.cc:75: Bind to role=worker, ip=11.11.72.4, port=34717, is_recovery=0
[04:19:57] src/van.cc:235: assign rank=9 to node role=worker, ip=11.11.72.4, port=34308, is_recovery=0
[04:19:57] src/van.cc:235: assign rank=11 to node role=worker, ip=11.11.72.4, port=34717, is_recovery=0
[04:19:57] src/van.cc:235: assign rank=8 to node role=server, ip=11.11.72.4, port=44464, is_recovery=0
[04:19:57] src/van.cc:235: assign rank=10 to node role=server, ip=11.11.72.4, port=44585, is_recovery=0
[04:19:57] src/van.cc:235: assign rank=12 to node role=server, ip=11.11.72.4, port=44899, is_recovery=0
[04:19:57] src/van.cc:235: assign rank=13 to node role=worker, ip=11.11.72.4, port=48209, is_recovery=0
[04:19:57] src/van.cc:235: assign rank=15 to node role=worker, ip=11.11.72.4, port=53854, is_recovery=0
[04:19:57] src/van.cc:235: assign rank=14 to node role=server, ip=11.11.72.4, port=56864, is_recovery=0
[04:19:57] src/van.cc:251: the scheduler is connected to 4 workers and 4 servers
[04:19:57] src/van.cc:[04:19:57] src/van.cc:281: W[15] is connected to others
281: W[13] is connected to others
[04:19:57] src/van.cc:281: W[11] is connected to others
[04:19:57] src/van.cc:281: W[9] is connected to others
[04:19:57] src/van.cc:281: S[14] is connected to others
[04:19:57] src/van.cc:291: Barrier count for 7 : 1
[04:19:57] src/van.cc:291: Barrier count for 7 : 2
[04:19:57] src/van.cc:291: Barrier count for 7 : 3
[04:19:57] src/van.cc:291: Barrier count for 7 : 4
[04:19:57] src/van.cc:291: Barrier count for 7 : 5
[04:19:57] src/van.cc:291: Barrier count for 7 : 6
```
Using gdb shows program hang on this line:https://github.com/dmlc/ps-lite/blob/5ac4af05e5048652154b80ca8f12b3d2751b50fa/src/postoffice.cc#L129
```
  barrier_cond_.wait(ulk, [this] {
      return barrier_done_;
    });
```
Any help will be appreciated

**Comment by zouweijia** [comment]
@feiyulv  Have you solved your problem? 

I am facing a similar situation:
When I submit a distributed job to a mpi-cluster, the program hangs after finishing last epoch training.

**Comment by chao1224** [comment]
@zouweijia 
I got the same problem. But sometimes before training, and sometimes after training.

**Comment by lzx1413** [comment]
same question, any ideas? thanks

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
Also, do please check out our [forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) for general "how-to" questions.

**Comment by lzhou-arch** [comment]
@feiyulv Have you figured it out? I have exactly the same issue. 

After I launch the job on a two-machine cluster, it hangs at 
[14:45:23] src/van.cc:183: Barrier count for 7 : 3

@szha Can you reopen this post since the issue is still pending? It would be great to have some pointers to debug the problem.

**Comment by jiajinyu** [comment]
We have encountered the same problem and found out the possible root cause. 

## Root cause:
A previous train  has nodes (worker/server) not killed for some reason. Then a new scheduler starts, again bind to the port, say, 9091, used by the scheduler in the previous train. The nodes in this train, and the nodes from previous instance now **both** send message to this scheduler. This confuses ZMQ because their **identity** frames are the same. So it might just drop message or do something weird. 

## Solution
Append a random string for each train in the identity string of the ZMQ identity frame (set in the `ZMQVan::Connet` in ps-lite. 

tagging people in this thread who shared same issue. Hope this works for you guys. 
cc @feiyulv @Jinwoo-Jeon @zouweijia @chao1224 @lzx1413 @lzhou-arch

**Comment by vandanavk** [comment]
@mxnet-label-bot add [Distributed]

**Comment by eric-haibin-lin** [comment]
@jiajinyu thanks for sharing the solution. Did you have time to create a PR for the fix?

**Comment by vrakesh** [comment]
@jiajinyu bump any update on PR for the fix?
